### PR TITLE
[SPARK-33848][SQL][FOLLOWUP] Introduce allowList for push into (if / case) branches 

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Introduce allowList push into (if / case) branches to fix potential bug.

### Why are the changes needed?

 Fix potential bug.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -548,41 +548,66 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
...
+  private def supportedUnaryExpression(e: UnaryExpression): Boolean = e match {
+    case _: IsNull | _: IsNotNull => true
+    case _: UnaryMathExpression | _: Abs | _: Bin | _: Factorial | _: Hex => true
+    case _: String2StringExpression | _: Ascii | _: Base64 | _: BitLength | _: Chr | _: Length =>
+      true
+    case _: CastBase => true
+    case _: GetDateField | _: LastDay => true
+    case _: ExtractIntervalPart => true
+    case _: ArraySetLike => true
+    case _ => false
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:561]: let's include `ExtractValue` as well, which is common with nested fields.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 2]
```diff
@@ -548,41 +548,66 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
     foldables.nonEmpty && others.length < 2
   }
 
+  // Not all UnaryExpression support push into (if / case) branches, e.g. Alias.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:551]: `Not all UnaryExpression can be pushed into (if / case) branches, e.g. Alias.`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 3]
```diff
@@ -548,41 +548,66 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
...
+    case _: String2StringExpression | _: Ascii | _: Base64 | _: BitLength | _: Chr | _: Length =>
+      true
+    case _: CastBase => true
+    case _: GetDateField | _: LastDay => true
+    case _: ExtractIntervalPart => true
+    case _: ArraySetLike => true
+    case _ => false
+  }
+
+  private def supportedBinaryExpression(e: BinaryExpression): Boolean = e match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:564]: let's add comments as well.

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
It's definitely not related to Spark R. I'm merging it to master, thanks!

**Comment by maropu** [comment]
Looks a safer approach and late lgtm.

side note: sorry for my late reply, but I think we'd be better to describe what's issue and how the PR fix it clearly in the PR description for better activity logs. At first I didn't understand what the potential bug was referring to and I found the previous discussion then: https://github.com/apache/spark/pull/30853#discussion_r549458098

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 4]
```diff
@@ -548,41 +548,68 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
...
+    case _: ExtractValue => true
+    case _ => false
+  }
+
+  // Not all BinaryExpression can be pushed into (if / case) branches.
+  private def supportedBinaryExpression(e: BinaryExpression): Boolean = e match {
+    case _: BinaryComparison | _: StringPredicate | _: StringRegexExpression => true
+    case _: BinaryArithmetic => true
+    case _: BinaryMathExpression => true
+    case _: AddMonths | _: DateAdd | _: DateAddInterval | _: DateDiff | _: DateSub => true
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:570]: What the property should the expression have to be here? For example, can I add `DateAddYMInterval`, `TimestampAddYMInterval` and `TimeAdd`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:570]: I think so

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:570]: ok. I opened the JIRA for that: SPARK-34841

**Pull request closed without merging**